cosi/protocol/sub_protocol.go

```diff
@@ -0,0 +1,279 @@
+
+	var secret kyber.Scalar
+
+	if p.IsRoot() {
+		// send commitment to super-protocol
+		if len(commitments) != 1 {
+			return fmt.Errorf("root node in subprotocol should have received 1 commitment,"+
+				"but received %d", len(commitments))
+		}
+		p.subCommitment <- commitments[0]
+	} else {
+		// do not commit if the verification does not succeed
+		if !<-verifyChan {
+			log.Lvl2(p.ServerIdentity().Address, "verification failed, terminating")
+			return nil
+		}
+
+		// otherwise, compute personal commitment and send to parent
+		var commitment kyber.Point
+		var mask *cosi.Mask
+		secret, commitment, mask, err = generateCommitmentAndAggregate(p.suite, p.TreeNodeInstance, p.Publics, commitments)
+		if err != nil {
+			return err
+		}
+		err = p.SendToParent(&Commitment{commitment, mask.Mask()})
+		if err != nil {
+			return err
+		}
+	}
+
+	// ----- Challenge -----
+	challenge, channelOpen := <-p.ChannelChallenge // from the leader
+	if !channelOpen {
+		return nil
+	}
+
+	log.Lvl3(p.ServerIdentity().Address, "received challenge")
+	if errs := p.Multicast(&challenge.Challenge, committedChildren...); len(errs) > 0 {
+		log.Lvl3(p.ServerIdentity().Address, "")
+	}
+
+	// ----- Response -----
+	if p.IsLeaf() {
+		p.ChannelResponse <- StructResponse{}
+	}
+	responses := make([]StructResponse, 0)
+
+	for _ = range committedChildren {
+		response, channelOpen := <-p.ChannelResponse
+		if !channelOpen {
+			return nil
+		}
+		responses = append(responses, response)
+	}
+	log.Lvl3(p.ServerIdentity().Address, "received all", len(responses), "response(s)")
+
+	if p.IsRoot() {
+		// send response to super-protocol
+		if len(responses) != 1 {
+			return fmt.Errorf(
+				"root node in subprotocol should have received 1 response, but received %v",
+				len(commitments))
+		}
+		p.subResponse <- responses[0]
+	} else {
+		// generate own response and send to parent
+		response, err := generateResponse(
+			p.suite, p.TreeNodeInstance, responses, secret, challenge.Challenge.CoSiChallenge)
+		if err != nil {
+			return err
+		}
+		err = p.SendToParent(&Response{response})
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+//HandleStop is called when a Stop message is send to this node.
+// It broadcasts the message and stops the node
+func (p *CoSiSubProtocolNode) HandleStop(stop StructStop) error {
+	defer p.Done()
+	if p.IsRoot() {
+		p.Broadcast(&stop.Stop)
+	}
+	return nil
+}
+
+// Start is done only by root and starts the subprotocol
+func (p *CoSiSubProtocolNode) Start() error {
+	log.Lvl3(p.ServerIdentity().Address, "Starting subCoSi")
+	if p.Msg == nil {
+		return errors.New("subprotocol does not have a proposal msg")
+	}
+	// p.Data can be nil
+	if p.Publics == nil || len(p.Publics) < 1 {
+		return errors.New("subprotocol has invalid public keys")
+	}
+	if p.verificationFn == nil {
+		return errors.New("subprotocol has an empty verification fn")
+	}
+	if p.Timeout < 10 {
```

Comment:
It seems like you want 10 * time.Second here?

Reply:
I wanted 10 nanoseconds, basically to make sure the timeout is realistic. I'll change it to 10 * time.Nanosecond to clarify.